libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Shrink.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE GADTs #-}
+
+module Test.Cardano.Ledger.Constrained.Shrink (
+  shrinkEnv,
+) where
+
+import Control.Monad
+import Data.Maybe
+
+import Cardano.Ledger.Core (Era (..))
+import Test.Cardano.Ledger.Constrained.Ast
+import Test.Cardano.Ledger.Constrained.Env
+import Test.Cardano.Ledger.Constrained.Monad
+import Test.Cardano.Ledger.Constrained.Rewrite
+import Test.Cardano.Ledger.Constrained.TypeRep
+
+-- | Shrink an environment subject to the constraints in the given dependency graph.
+--   The strategy is
+--    * pick a variable
+--    * shrink its value to something that still satisfies the defining constraints for the variable
+--    * adjust the values for variables depending on the picked variables to fix constraint
+--      violations resulting from the changed value
+shrinkEnv :: Era era => DependGraph era -> Env era -> [Env era]
+shrinkEnv (DependGraph vs) env =
+  [ env'
+  | (before, (x, cs), after) <- splits vs
+  , env' <- shrinkOneVar env (map fst before) x cs after
+  ]
+
+splits :: [a] -> [([a], a, [a])]
+splits [] = []
+splits (x : xs) = ([], x, xs) : [(x : ys, y, zs) | (ys, y, zs) <- splits xs]
```

Comment:
A comment about how this is to be used. I assume it is applied to the dependency graph?

Reply:
It's only used in the function above. I made it local.